libnavigation-core/src/main/java/com/mapbox/navigation/core/routeoptions/RouteOptionsUpdater.kt

```diff
@@ -255,13 +321,28 @@ class RouteOptionsUpdater {
         oldMap: Map<String, JsonElement>,
         key: String,
         nextCoordinateIndex: Int,
+        serverAddedWaypointParametersWithIndex: List<Pair<Int, String>>,
+        waypointsCount: Int
     ) {
-        if (key !in oldMap) {
+        if (key !in oldMap && serverAddedWaypointParametersWithIndex.isEmpty()) {
             return
         }
-        val oldValue = oldMap[key]!!.asString
-        val elements = oldValue.split(";")
-        val newValue = ";" + elements.drop(nextCoordinateIndex).joinToString(";")
+        val elements: List<String> = oldMap[key]?.asString?.let {
+            it.split(";").toMutableList().apply {
+                serverAddedWaypointParametersWithIndex.forEach { (index, param) ->
+                    this.add(index, param)
+                }
+            }
+        } ?: MutableList(waypointsCount) { "" }.apply {
+            serverAddedWaypointParametersWithIndex.forEach { (index, param) ->
+                this[index] = param
+            }
+        }
+        val remainingElements = elements.drop(nextCoordinateIndex)
+        if (remainingElements.all { it.isEmpty() }) {
```

Comment:
I don't think this is right. Imagine the situation. For some key we had: "key=1;2;;;;;", we passed 2 waypoints. Now this line will evaluate to true and we won't change anything. As a result, we'll get "key=1;2;;;;;", but we should either remove the key or update it to "key=;;;;;".

Reply:
> For some key we had: "key=1;2;;;;;", we passed 2 waypoints. Now this line will evaluate to true and we won't change anything

In this case key will be filtered out for the next request. You can try removing this if and run `EvRouteOptionsUpdaterTest`

Reply:
Discussed internally.